indigo/analysis/refs/provisions.py

```diff
@@ -0,0 +1,419 @@
+import re
+from dataclasses import dataclass
+from typing import List, Union, Tuple
+
+from lxml import etree
+from lxml.etree import Element
+
+from cobalt import FrbrUri
+from docpipe.matchers import TextPatternMatcher
+from docpipe.xmlutils import wrap_text
+from indigo.plugins import LocaleBasedMatcher, plugins
+from cobalt.schemas import AkomaNtoso30
+from indigo.analysis.refs.provision_refs import parse, ParseError
+
+
+@dataclass
+class ProvisionRef:
+    """A single provision reference.
+
+    For "Section 32(a)(i) and (b)", this represents "(a)(i)".
+    """
+    # text of the matched reference
+    text: str
+    # start and end positions of the matched reference
+    start_pos: int
+    end_pos: int
+    # "and_or", "range"
+    separator: Union[str, None] = None
+    # the next ref in the chain, eg "(i)" for "(a)(i)"
+    child: 'ProvisionRef' = None
+    element: Union[Element, None] = None
+    # eId of the matched element
+    eId: Union[str, None] = None
+
+
+@dataclass
+class MainProvisionRef:
+    """The root of a series of references.
+
+    For "Section 32(a) and (b)", this represents "Section 32(a)".
+    - name: Section
+    - ref: ProvisionRef("32", ..., ProvisionRef("(a)", ...))
+    - sub_refs: ProvisionRef("(b)", ...)
+    """
+    # eg "section" or "part"
+    name: str
+    # the primary ref
+    ref: ProvisionRef
+    # secondary refs that are resolved relative to the primary
+    sub_refs: List[ProvisionRef] = None
```

Comment:
something complained about me using an object in a list like this the other day; is that because I didn't use `List` from `typing`?

Reply:
> something complained about me using an object in a list like this the other day; is that because I didn't use `List` from `typing`?

Probably